File: chroma-league/src/main/java/net/booone/chromaleague/hud/parts/health/LostHealthAnimation.java
package net.booone.chromaleague.hud.parts.health;

import net.booone.chromaleague.hud.parts.Background;
import net.booone.razersdk.animation.AnimatedFrame;
import net.booone.razersdk.animation.SimpleFrame;
import net.booone.razersdk.color.StaticColor;
import net.booone.razersdk.color.TransitionColor;
import net.booone.razersdk.sdk.RzKey;

import java.util.List;
import java.util.stream.IntStream;

public class LostHealthAnimation extends AnimatedFrame {
    private static final int LOST_HEALTH_CHANGE_STEPS = 5;

    public LostHealthAnimation(double previousHp, double currentHp) {
        createLostHealthAnimation(previousHp, currentHp);
    }

    private void createLostHealthAnimation(double previousHp, double currentHp) {
        final List<RzKey> rzKeys = computeLostHealth(previousHp, currentHp);
        if (!rzKeys.isEmpty()) {
            final TransitionColor transitionColor = new TransitionColor(StaticColor.RED, Background.DEFAULT_BACKGROUND_COLOR, LOST_HEALTH_CHANGE_STEPS);
            IntStream.range(0, LOST_HEALTH_CHANGE_STEPS).forEach(i -> addAnimationFrame(new SimpleFrame(rzKeys, transitionColor.getColor())));
        }
    }

    private static List<RzKey> computeLostHealth(double previousHp, double currentHp) {
        return HealthBar.getHpBarPart(currentHp, previousHp);
    }
}

File: src/hud/parts/health/LostHealthAnimation.ts
import {AnimatedFrame} from '../../../razer-sdk/animation/AnimatedFrame.js';
import {SimpleFrame} from '../../../razer-sdk/animation/SimpleFrame.js';
import {StaticColor} from '../../../razer-sdk/color/StaticColor.js';
import {TransitionColor} from '../../../razer-sdk/color/TransitionColor.js';
import {RzKey} from '../../../razer-sdk/sdk/RzKey.js';
import {DEFAULT_BACKGROUND_COLOR} from '../Background.js';
import {getHpBarPart} from './HealthBar.js';

const LOST_HEALTH_CHANGE_STEPS = 5;

export class LostHealthAnimation extends AnimatedFrame {
    constructor(previousHp: number, currentHp: number) {
        super();
        this.createLostHealthAnimation(previousHp, currentHp);
    }

    private createLostHealthAnimation(previousHp: number, currentHp: number): void {
        const rzKeys = computeLostHealth(previousHp, currentHp);
        if (rzKeys.length > 0) {
            const transitionColor = new TransitionColor(StaticColor.RED, DEFAULT_BACKGROUND_COLOR, LOST_HEALTH_CHANGE_STEPS);
            for (let i = 0; i < LOST_HEALTH_CHANGE_STEPS; i++) {
                this.addAnimationFrame(new SimpleFrame(rzKeys, transitionColor.getColor()));
            }
        }
    }
}

function computeLostHealth(previousHp: number, currentHp: number): RzKey[] {
    return getHpBarPart(currentHp, previousHp);
}
